### PR TITLE
Editor: Add a featured image dropzone above the post title

### PR DESCRIPTION
**Pull request by iandunn**
The old editor had the ability to drop a featured image above the post title, but that's missing from the current editor. This PR restores that ability.

In order to do that, the existing `MediaLibraryDropZone` has to be adjusted to allow room for the new `FeaturedImageDropZone`.

This currently works, but `dropZoneIsActive` is currently hardcoded, because I'm not sure what the best way to determine it would be. The overall approach may also need some refinement, or a change in direction.

Also, the adjustments made to the `MediaLibraryDropZone` introduce two unintended consequences that may need to be fixed before this is ready for merge:

1. It's now more likely that a user will drop a file into an area outside a dropzone, since we no longer have a large zone covering almost the entire screen. If they do this, then the browser will open the file as a new page in the history, taking them away from Calypso.
1. The TinyMCE toolbar pinning/unpinning breaks when the dropzones are active. It's unlikely the user will try to scroll while dragging-dropping a file, but it's worth noting regardless.

See #13277

### CODE REVIEW

File: client/post-editor/editor-featured-image/index.jsx [hunk 1]
```diff
@@ -121,8 +124,10 @@ class EditorFeaturedImage extends Component {
 	render() {
 		const { site, post } = this.props;
 		const featuredImageId = getFeaturedImageId( post );
+		const dropZoneIsActive = true;  // todo not sure best way to determine if it's active w/ out coupling. maybe create a selector to pick isDraggingOverDocument from the redux store?
```
Comment (bisko) [inline client/post-editor/editor-featured-image/index.jsx]: I think you can leverage the already [existing selector](https://github.com/Automattic/wp-calypso/blob/master/client/state/selectors/is-drop-zone-visible.js#L6).

File: client/post-editor/post-editor.jsx [hunk 2]
```diff
@@ -380,6 +381,16 @@ export const PostEditor = React.createClass( {
 								onBlur={ this.copySelectedText }
 								onTextEditorChange={ this.onEditorContentChange } />
 						</div>
+
+						{/* todo need some code somewhere -- maybe in here -- so that if user drops a file on an empty area (an area that isn't a dropzone),
```
Comment (bisko) [inline client/post-editor/post-editor.jsx]: I think this would be awesome to have as a separate PR, so we can keep this one more on-point with the issue at hand. Also it will decouple the features and allow us to merge the false drop update before merging this PR.

File: client/post-editor/editor-drawer/featured-image.jsx [hunk 3]
```diff
@@ -64,6 +64,7 @@ class EditorDrawerFeaturedImage extends Component {
 						site={ site }
 						post={ post }
 					/>
+					{/* todo maybe change this to use dropzone as well? why would the dropzone be attached to the EditorDrawerWell instead of the FeaturedImage? */}
```
Comment (bisko) [inline client/post-editor/editor-drawer/featured-image.jsx]: I think I added it here because there was a styling problem, i.e. the DropZone would only stretch over the middle area, without covering the whole "block". 

It can definitely be improved and it would be awesome if we can get this in as a separate, refactor PR!

File: client/post-editor/post-editor.jsx (hunk 2 above)
Comment (iandunn) [inline client/post-editor/post-editor.jsx]: That's a great idea, thanks!

File: client/post-editor/editor-featured-image/index.jsx (hunk 1 above)
Comment (iandunn) [inline client/post-editor/editor-featured-image/index.jsx]: Ah, that didn't occur to me but that sounds like the way to go, thanks!

File: client/components/tinymce/index.jsx [hunk 4]
```diff
@@ -496,7 +496,7 @@ module.exports = React.createClass( {
 		} );
 
 		return (
-			<div>
+			<div className="tinymce-container">
```
Comment (iandunn) [inline client/components/tinymce/index.jsx]: @bisko this caused an eslint error because `jsx-class-namespace` expects the `className` to be `tinymce`, not `tinymce-container`.

I used `tinymce-container`, though, because `tinymce` is already used on the `<textarea>`. Would the best thing to do here be to make a new PR to rename the `<textarea>` to something like `tinymce-raw-text`?

**Comment by lancewillett** [comment]
Once this is ready for further testing, and to help communicate "in 1,000 words" — aka visual oxygen — could you include a movie or screenshot to illustrate what's expected to happen? Thank you.

**Comment by iandunn** [comment]
Good point, thanks!

I uploaded a movie to https://cloudup.com/cVJB8CuKEwZ

**Comment by alisterscott** [comment]
I tested this and it works well. One suggestion I would make is to add a 'remove' icon (X) similar to the one under the featured image section in the sidebar so if someone inadvertently added an image as a featured image by dropping it here instead of the editor they could quickly remove it. There's currently no way to do this for the image above the editor at the moment (only a way to change it)

screen shot 2017-08-30 at 12 56 40 pm

**Comment by iandunn** [comment]
Good catch, I'll add that :)

**Comment by iandunn** [comment]
The remove button has been added.

**Comment by alisterscott** [comment]
Deleting featured image from above the editor works well - thank you

File: client/components/tinymce/index.jsx [hunk 5]
```diff
@@ -491,8 +492,18 @@ module.exports = React.createClass( {
...
 
+		/*
+		 * Using `classnames()` here is just a hack to avoid the linter complaining that the
+		 * container is named `tinymce-container` instead of `tinymce`. Ideally the containing
+		 * `div` and the `textarea` should be refactored so that the `div` has the `tinymce`
+		 * class, but that would interfere with higher priority fixes. This component is slated
+		 * for some refactoring in the near future, so that will be a more convenient time to
+		 * clean this up.
+		 */
+		const containerClassName = classnames( { 'tinymce-container': true } );
```
Comment (jblz) [inline client/components/tinymce/index.jsx]: This can just be `classnames( 'tinymce-container' )`

File: client/post-editor/editor-featured-image/index.jsx [hunk 6]
```diff
@@ -122,7 +139,8 @@ class EditorFeaturedImage extends Component {
 		const { site, post } = this.props;
 		const featuredImageId = getFeaturedImageId( post );
 		const classes = classnames( 'editor-featured-image', {
-			'is-assigned': !! PostUtils.getFeaturedImageId( this.props.post )
+			'is-assigned': !! PostUtils.getFeaturedImageId( this.props.post ),
```
Comment (jblz) [inline client/post-editor/editor-featured-image/index.jsx]: I _think_ you can drop the `!!` here since it just does a "truthiness" check.

File: client/components/remove-button/index.jsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import { noop, identity } from 'lodash';
+import Gridicon from 'gridicons';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+
+class RemoveButton extends Component {
```
Comment (jblz) [inline client/components/remove-button/index.jsx]: Nice :) I like this refactor.

File: client/post-editor/editor-featured-image/index.jsx (hunk 6 above)
Comment (iandunn) [reply client/post-editor/editor-featured-image/index.jsx]: Yeah, it works the same in my tests.

**Comment by iandunn** [comment]
I'm not able to reproduce the loading order issue on an older Macbook, either in FF or Chrome. I'll have access to a faster iMac next week that I can try on, though.

In the meantime, I've created #18061 and #18064, in order to break this PR into smaller pieces.

**Comment by bisko** [comment]
@jblz :
> I'm not able to reproduce this on master, so it looks like this issue might stem from these changes after all.

Can you check if the same issue happens with the Featured Image drop zone in the sidebar? Both on this branch and master?

**Comment by iandunn** [comment]
This PR was getting too big, so I've split some of it off into #18061 and #18064. The first one is merged, and this has been rebased against `master`. 

I'm still working on the second one, but once it's merged and this is rebased again, this branch should be much more manageable.

**Comment by iandunn** [comment]
Quick notes-to-self for later, regarding the drop-file-before-js-loaded issue:

* the issue is probably easy to reproduce using the network throttle feature in Chrome dev tools
* one solution might be to load a global drop handler before most other JS, so that the browser doesn't navigate to dropped files outside of a dropzone. that would fix this, plus other issues, like a user missing the dropzone.
* this probably isn't a blocker

**Comment by iandunn** [comment]
With dev tool throttling, I was able to reproduce the problem with dropping images before scripts have loaded, even on `master`. So, I don't think it was introduced here. 

I think @marekhrabe's suggestion about a global file handler is a good idea, though, since it would cover this and other situations. I opened #18463 to file that.

File: client/post-editor/editor-featured-image/index.jsx [hunk 8]
```diff
@@ -159,6 +166,8 @@ class EditorFeaturedImage extends Component {
 					</Button>
 					{ featuredImageId ? <RemoveButton onRemove={ EditorFeaturedImage.removeImage } /> : '' }
 				</div>
+
+				{ this.props.hasDropZone ? <FeaturedImageDropZone /> : '' }
```
Comment (jblz) [inline client/post-editor/editor-featured-image/index.jsx]: You should be able to use the conventional:
`this.props.hasDropZone && <FeaturedImageDropZone />;`
...instead of the ternary here as null values should be ignored.

Comment (marekhrabe) [reply client/post-editor/editor-featured-image/index.jsx]: Even when keeping this construct, `null` would be preferable over `""` an empty string

Comment (iandunn) [reply client/post-editor/editor-featured-image/index.jsx]: Ah, good tips, I'll incorporate that before rebasing. Thanks!
